File: projects/basis/sdk-review/py-v2/src/modules/Vesting.ts
import { BasisClient } from '../BasisClient';
import AVestingArtifact from '../abis/A_VestingContract.json';
import IERC20Artifact from '../abis/IERC20.json';
import { Address } from 'viem';

export class VestingModule {
  private client: BasisClient;
  private vestingAddress: Address;

  constructor(client: BasisClient, vestingAddress: Address) {
    this.client = client;
    this.vestingAddress = vestingAddress;
  }

  private async _syncTx(txHash: string) {
    try {
      await this.client.api.syncTransaction(txHash);
    } catch (e: any) {
      console.warn('Sync warning:', e.message || e);
    }
  }

  private async approveIfNeeded(tokenAddress: Address, spender: Address, amount: bigint) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Wallet account is required for approval.");
    }
    const account = this.client.walletClient.account;
    const allowance = await this.client.publicClient.readContract({
      address: tokenAddress,
      abi: IERC20Artifact.abi,
      functionName: 'allowance',
      args: [account.address, spender],
    }) as bigint;

    if (allowance < amount) {
      const { request } = await this.client.publicClient.simulateContract({
        account,
        address: tokenAddress,
        abi: IERC20Artifact.abi,
        functionName: 'approve',
        args: [spender, amount],
      });
      const hash = await this.client.walletClient.writeContract(request);
      await this.client.publicClient.waitForTransactionReceipt({ hash });
    }
  }

  private async getFeeAmount(): Promise<bigint> {
    try {
      return await this.client.publicClient.readContract({
        address: this.vestingAddress,
        abi: AVestingArtifact.abi,
        functionName: 'feeAmount',
      }) as bigint;
    } catch {
      return 0n;
    }
  }

  /**
   * Creates a gradual vesting schedule.
   * Auto-approves the token to the vesting contract and attaches the creation fee.
   */
  async createGradualVesting(
    beneficiary: Address,
    token: Address,
    totalAmount: bigint,
    startTime: bigint,
    durationInDays: bigint,
    timeUnit: number,
    memo: string,
    ecosystem: Address
  ) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    // Auto-approve token to the vesting contract
    await this.approveIfNeeded(token, this.vestingAddress, totalAmount);

    // Fetch fee
    const feeAmount = await this.getFeeAmount();

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.vestingAddress,
      abi: AVestingArtifact.abi,
      functionName: 'createGradualVesting',
      args: [beneficiary, token, totalAmount, startTime, durationInDays, timeUnit, memo, ecosystem],
      value: feeAmount,
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    this._syncTx(hash);

    return { hash, receipt };
  }

  /**
   * Creates a cliff vesting schedule.
   */
  async createCliffVesting(
    beneficiary: Address,
    token: Address,
    totalAmount: bigint,
    unlockTime: bigint,
    memo: string,
    ecosystem: Address
  ) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    await this.approveIfNeeded(token, this.vestingAddress, totalAmount);
    const feeAmount = await this.getFeeAmount();

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.vestingAddress,
      abi: AVestingArtifact.abi,
      functionName: 'createCliffVesting',
      args: [beneficiary, token, totalAmount, unlockTime, memo, ecosystem],
      value: feeAmount,
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    this._syncTx(hash);

    return { hash, receipt };
  }

  /**
   * Claims unlocked tokens.
   */
  async claimTokens(vestingId: bigint) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.vestingAddress,
      abi: AVestingArtifact.abi,
      functionName: 'claimTokens',
      args: [vestingId],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    this._syncTx(hash);

    return { hash, receipt };
  }

  /**
   * Leverages locked tokens for a loan.
   */
  async takeLoanOnVesting(vestingId: bigint) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.vestingAddress,
      abi: AVestingArtifact.abi,
      functionName: 'takeLoanOnVesting',
      args: [vestingId],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    this._syncTx(hash);

    return { hash, receipt };
  }

  /**
   * Repays a loan taken on a vesting schedule.
   * Auto-approves the borrowed token (USDB) to the vesting contract.
   */
  async repayLoanOnVesting(vestingId: bigint) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    // Approve USDB (the borrowed token) to the vesting contract for repayment
    const usdbBalance = await this.client.publicClient.readContract({
      address: this.client.usdbAddress,
      abi: IERC20Artifact.abi,
      functionName: 'balanceOf',
      args: [this.client.walletClient.account.address],
    }) as bigint;
    if (usdbBalance > 0n) {
      await this.approveIfNeeded(this.client.usdbAddress, this.vestingAddress, usdbBalance);
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.vestingAddress,
      abi: AVestingArtifact.abi,
      functionName: 'repayLoanOnVesting',
      args: [vestingId],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    this._syncTx(hash);

    return { hash, receipt };
  }

  /**
   * Gets details of a specific vesting schedule.
   */
  async getVestingDetails(vestingId: bigint) {
    return this.client.publicClient.readContract({
      address: this.vestingAddress,
      abi: AVestingArtifact.abi,
      functionName: 'getVestingDetails',
      args: [vestingId],
    });
  }

  /**
   * Gets the current claimable amount for a vesting schedule.
   */
  async getClaimableAmount(vestingId: bigint) {
    return this.client.publicClient.readContract({
      address: this.vestingAddress,
      abi: AVestingArtifact.abi,
      functionName: 'getClaimableAmount',
      args: [vestingId],
    });
  }

  /**
   * Creates gradual vesting schedules for multiple beneficiaries in a single transaction.
   * Auto-approves the sum of all amounts and attaches the creation fee.
   */
  async batchCreateGradualVesting(
    beneficiaries: Address[],
    token: Address,
    totalAmounts: bigint[],
    userMemos: string[],
    startTime: bigint,
    durationInDays: bigint,
    timeUnit: number,
    ecosystem: Address
  ) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    // Sum all amounts for approval
    const totalApproval = totalAmounts.reduce((sum, amt) => sum + amt, 0n);
    await this.approveIfNeeded(token, this.vestingAddress, totalApproval);

    const feeAmount = await this.getFeeAmount();

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.vestingAddress,
      abi: AVestingArtifact.abi,
      functionName: 'batchCreateGradualVesting',
      args: [beneficiaries, token, totalAmounts, userMemos, startTime, durationInDays, timeUnit, ecosystem],
      value: feeAmount,
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    this._syncTx(hash);

    return { hash, receipt };
  }

  /**
   * Creates cliff vesting schedules for multiple beneficiaries in a single transaction.
   * Auto-approves the sum of all amounts and attaches the creation fee.
   */
  async batchCreateCliffVesting(
    beneficiaries: Address[],
    token: Address,
    totalAmounts: bigint[],
    unlockTime: bigint,
    userMemos: string[],
    ecosystem: Address
  ) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    const totalApproval = totalAmounts.reduce((sum, amt) => sum + amt, 0n);
    await this.approveIfNeeded(token, this.vestingAddress, totalApproval);

    const feeAmount = await this.getFeeAmount();

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.vestingAddress,
      abi: AVestingArtifact.abi,
      functionName: 'batchCreateCliffVesting',
      args: [beneficiaries, token, totalAmounts, unlockTime, userMemos, ecosystem],
      value: feeAmount,
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    this._syncTx(hash);

    return { hash, receipt };
  }

  /**
   * Changes the beneficiary of a vesting schedule.
   */
  async changeBeneficiary(vestingId: bigint, newBeneficiary: Address) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.vestingAddress,
      abi: AVestingArtifact.abi,
      functionName: 'changeBeneficiary',
      args: [vestingId, newBeneficiary],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    this._syncTx(hash);

    return { hash, receipt };
  }

  /**
   * Extends the vesting period by additional days.
   */
  async extendVestingPeriod(vestingId: bigint, additionalDays: bigint) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.vestingAddress,
      abi: AVestingArtifact.abi,
      functionName: 'extendVestingPeriod',
      args: [vestingId, additionalDays],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    this._syncTx(hash);

    return { hash, receipt };
  }

  /**
   * Adds more tokens to an existing vesting schedule.
   * Auto-approves the token to the vesting contract.
   */
  async addTokensToVesting(vestingId: bigint, additionalAmount: bigint) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    // Read vesting details to get the token address
    const details = await this.getVestingDetails(vestingId) as any;
    const token = details.token ?? details[2];
    await this.approveIfNeeded(token, this.vestingAddress, additionalAmount);

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.vestingAddress,
      abi: AVestingArtifact.abi,
      functionName: 'addTokensToVesting',
      args: [vestingId, additionalAmount],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    this._syncTx(hash);

    return { hash, receipt };
  }

  /**
   * Transfers the creator role of a vesting schedule to a new address.
   */
  async transferCreatorRole(vestingId: bigint, newCreator: Address) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.vestingAddress,
      abi: AVestingArtifact.abi,
      functionName: 'transferCreatorRole',
      args: [vestingId, newCreator],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    this._syncTx(hash);

    return { hash, receipt };
  }

  /**
   * Returns all vesting IDs for a given beneficiary.
   */
  async getVestingsByBeneficiary(beneficiary: Address): Promise<bigint[]> {
    return this.client.publicClient.readContract({
      address: this.vestingAddress,
      abi: AVestingArtifact.abi,
      functionName: 'getVestingsByBeneficiary',
      args: [beneficiary],
    }) as Promise<bigint[]>;
  }

  /**
   * Returns the total vested amount for a vesting schedule.
   */
  async getVestedAmount(vestingId: bigint): Promise<bigint> {
    return this.client.publicClient.readContract({
      address: this.vestingAddress,
      abi: AVestingArtifact.abi,
      functionName: 'getVestedAmount',
      args: [vestingId],
    }) as Promise<bigint>;
  }

  /**
   * Returns the active loan ID for a vesting schedule.
   */
  async getActiveLoan(vestingId: bigint): Promise<bigint> {
    return this.client.publicClient.readContract({
      address: this.vestingAddress,
      abi: AVestingArtifact.abi,
      functionName: 'getActiveLoan',
      args: [vestingId],
    }) as Promise<bigint>;
  }

  /**
   * Returns vesting IDs for a given token within a specified index range.
   */
  async getTokenVestingIds(token: Address, startIndex: bigint, endIndex: bigint): Promise<bigint[]> {
    return this.client.publicClient.readContract({
      address: this.vestingAddress,
      abi: AVestingArtifact.abi,
      functionName: 'getTokenVestingIds',
      args: [token, startIndex, endIndex],
    }) as Promise<bigint[]>;
  }

  /**
   * Returns vesting details for multiple vesting IDs in a single call.
   */
  async getVestingDetailsBatch(vestingIds: bigint[]) {
    return this.client.publicClient.readContract({
      address: this.vestingAddress,
      abi: AVestingArtifact.abi,
      functionName: 'getVestingDetailsBatch',
      args: [vestingIds],
    });
  }

  /**
   * Returns the total number of vesting schedules created.
   */
  async getVestingCount(): Promise<bigint> {
    return this.client.publicClient.readContract({
      address: this.vestingAddress,
      abi: AVestingArtifact.abi,
      functionName: 'vestingCount',
    }) as Promise<bigint>;
  }

  /**
   * Returns all vesting IDs created by a given creator.
   */
  async getVestingsByCreator(creator: Address): Promise<bigint[]> {
    return this.client.publicClient.readContract({
      address: this.vestingAddress,
      abi: AVestingArtifact.abi,
      functionName: 'getVestingsByCreator',
      args: [creator],
    }) as Promise<bigint[]>;
  }
}

File: projects/basis/sdk-review/py-v2/basis/modules/vesting.py
import logging

from web3 import Web3
from .factory import load_abi

logger = logging.getLogger(__name__)


class VestingModule:
    def __init__(self, client, vesting_address: str):
        self.client = client
        self.vesting_address = Web3.to_checksum_address(vesting_address)
        self.vesting_abi = load_abi('A_VestingContract.json')
        self.erc20_abi = load_abi('IERC20.json')
        self.contract = self.client.web3.eth.contract(address=self.vesting_address, abi=self.vesting_abi)

    def _approve_if_needed(self, token_address: str, spender: str, amount: int):
        if not self.client.account:
            raise ValueError("Wallet account is required for approval.")
        checksum_token = Web3.to_checksum_address(token_address)
        checksum_spender = Web3.to_checksum_address(spender)
        token_contract = self.client.web3.eth.contract(address=checksum_token, abi=self.erc20_abi)
        allowance = token_contract.functions.allowance(
            self.client.account.address, checksum_spender
        ).call()
        if allowance < amount:
            func = token_contract.functions.approve(checksum_spender, amount)
            tx = func.build_transaction({
                'from': self.client.account.address,
                'nonce': self.client.web3.eth.get_transaction_count(self.client.account.address),
            })
            signed_tx = self.client.web3.eth.account.sign_transaction(tx, private_key=self.client.account.key)
            tx_hash = self.client.web3.eth.send_raw_transaction(signed_tx.raw_transaction)
            self.client.web3.eth.wait_for_transaction_receipt(tx_hash)

    def _get_fee_amount(self) -> int:
        try:
            return self.contract.functions.feeAmount().call()
        except Exception:
            return 0

    def _build_and_send_tx(self, function_call, value=0):
        if not self.client.account:
            raise ValueError("Stateful initialization (private_key) is required for write methods.")

        tx = function_call.build_transaction({
            'from': self.client.account.address,
            'nonce': self.client.web3.eth.get_transaction_count(self.client.account.address),
            'value': value,
        })

        signed_tx = self.client.web3.eth.account.sign_transaction(tx, private_key=self.client.account.key)
        tx_hash = self.client.web3.eth.send_raw_transaction(signed_tx.raw_transaction)
        receipt = self.client.web3.eth.wait_for_transaction_receipt(tx_hash)

        return {
            'hash': tx_hash.hex(),
            'receipt': receipt
        }

    def _sync_tx(self, tx_hash: str):
        """Sync tx to backend. Non-fatal on failure."""
        try:
            if not tx_hash.startswith("0x"):
                tx_hash = "0x" + tx_hash
            self.client.api.sync_transaction(tx_hash)
        except Exception as e:
            logger.warning("Sync warning: %s", e)

    def create_gradual_vesting(self, beneficiary: str, token: str, total_amount: int, start_time: int, duration_in_days: int, time_unit: int, memo: str, ecosystem: str):
        """Creates a gradual vesting. Auto-approves token and attaches fee."""
        self._approve_if_needed(token, self.vesting_address, total_amount)
        fee = self._get_fee_amount()
        func = self.contract.functions.createGradualVesting(
            Web3.to_checksum_address(beneficiary),
            Web3.to_checksum_address(token),
            total_amount, start_time, duration_in_days, time_unit, memo,
            Web3.to_checksum_address(ecosystem)
        )
        result = self._build_and_send_tx(func, value=fee)
        self._sync_tx(result['hash'])
        return result

    def create_cliff_vesting(self, beneficiary: str, token: str, total_amount: int, unlock_time: int, memo: str, ecosystem: str):
        """Creates a cliff vesting. Auto-approves token and attaches fee."""
        self._approve_if_needed(token, self.vesting_address, total_amount)
        fee = self._get_fee_amount()
        func = self.contract.functions.createCliffVesting(
            Web3.to_checksum_address(beneficiary),
            Web3.to_checksum_address(token),
            total_amount, unlock_time, memo,
            Web3.to_checksum_address(ecosystem)
        )
        result = self._build_and_send_tx(func, value=fee)
        self._sync_tx(result['hash'])
        return result

    def claim_tokens(self, vesting_id: int):
        func = self.contract.functions.claimTokens(vesting_id)
        result = self._build_and_send_tx(func)
        self._sync_tx(result['hash'])
        return result

    def take_loan_on_vesting(self, vesting_id: int):
        func = self.contract.functions.takeLoanOnVesting(vesting_id)
        result = self._build_and_send_tx(func)
        self._sync_tx(result['hash'])
        return result

    def repay_loan_on_vesting(self, vesting_id: int):
        """Repays a loan on vesting. Auto-approves USDB to vesting contract."""
        if not self.client.account:
            raise ValueError("Stateful initialization (private_key) is required.")
        usdb_contract = self.client.web3.eth.contract(
            address=Web3.to_checksum_address(self.client.usdb_address), abi=self.erc20_abi
        )
        balance = usdb_contract.functions.balanceOf(self.client.account.address).call()
        if balance > 0:
            self._approve_if_needed(self.client.usdb_address, self.vesting_address, balance)
        func = self.contract.functions.repayLoanOnVesting(vesting_id)
        result = self._build_and_send_tx(func)
        self._sync_tx(result['hash'])
        return result

    def get_vesting_details(self, vesting_id: int):
        return self.contract.functions.getVestingDetails(vesting_id).call()

    def get_claimable_amount(self, vesting_id: int):
        return self.contract.functions.getClaimableAmount(vesting_id).call()

    def batch_create_gradual_vesting(self, beneficiaries: list[str], token: str, total_amounts: list[int], user_memos: list[str], start_time: int, duration_in_days: int, time_unit: int, ecosystem: str):
        """Creates gradual vestings for multiple beneficiaries. Auto-approves sum of amounts and attaches fee."""
        checksum_beneficiaries = [Web3.to_checksum_address(b) for b in beneficiaries]
        checksum_token = Web3.to_checksum_address(token)
        checksum_ecosystem = Web3.to_checksum_address(ecosystem)
        total = sum(total_amounts)
        self._approve_if_needed(token, self.vesting_address, total)
        fee = self._get_fee_amount()
        func = self.contract.functions.batchCreateGradualVesting(
            checksum_beneficiaries, checksum_token, total_amounts, user_memos,
            start_time, duration_in_days, time_unit, checksum_ecosystem
        )
        result = self._build_and_send_tx(func, value=fee)
        self._sync_tx(result['hash'])
        return result

    def batch_create_cliff_vesting(self, beneficiaries: list[str], token: str, total_amounts: list[int], unlock_time: int, user_memos: list[str], ecosystem: str):
        """Creates cliff vestings for multiple beneficiaries. Auto-approves sum of amounts and attaches fee."""
        checksum_beneficiaries = [Web3.to_checksum_address(b) for b in beneficiaries]
        checksum_token = Web3.to_checksum_address(token)
        checksum_ecosystem = Web3.to_checksum_address(ecosystem)
        total = sum(total_amounts)
        self._approve_if_needed(token, self.vesting_address, total)
        fee = self._get_fee_amount()
        func = self.contract.functions.batchCreateCliffVesting(
            checksum_beneficiaries, checksum_token, total_amounts, unlock_time,
            user_memos, checksum_ecosystem
        )
        result = self._build_and_send_tx(func, value=fee)
        self._sync_tx(result['hash'])
        return result

    def change_beneficiary(self, vesting_id: int, new_beneficiary: str):
        """Changes the beneficiary of a vesting."""
        func = self.contract.functions.changeBeneficiary(vesting_id, Web3.to_checksum_address(new_beneficiary))
        result = self._build_and_send_tx(func)
        self._sync_tx(result['hash'])
        return result

    def extend_vesting_period(self, vesting_id: int, additional_days: int):
        """Extends the vesting period by additional days."""
        func = self.contract.functions.extendVestingPeriod(vesting_id, additional_days)
        result = self._build_and_send_tx(func)
        self._sync_tx(result['hash'])
        return result

    def add_tokens_to_vesting(self, vesting_id: int, additional_amount: int):
        """Adds tokens to an existing vesting. Reads vesting details to get token, then approves."""
        details = self.get_vesting_details(vesting_id)
        token = details[2]  # token address from vesting details (creator=0, beneficiary=1, token=2)
        self._approve_if_needed(token, self.vesting_address, additional_amount)
        func = self.contract.functions.addTokensToVesting(vesting_id, additional_amount)
        result = self._build_and_send_tx(func)
        self._sync_tx(result['hash'])
        return result

    def transfer_creator_role(self, vesting_id: int, new_creator: str):
        """Transfers the creator role of a vesting to a new address."""
        func = self.contract.functions.transferCreatorRole(vesting_id, Web3.to_checksum_address(new_creator))
        result = self._build_and_send_tx(func)
        self._sync_tx(result['hash'])
        return result

    def get_vested_amount(self, vesting_id: int) -> int:
        """Returns the total vested amount for a vesting schedule."""
        return self.contract.functions.getVestedAmount(vesting_id).call()

    def get_active_loan(self, vesting_id: int) -> int:
        """Returns the active loan ID for a vesting schedule."""
        return self.contract.functions.getActiveLoan(vesting_id).call()

    def get_token_vesting_ids(self, token: str, start_index: int, end_index: int) -> list:
        """Returns vesting IDs for a given token within a specified index range."""
        return self.contract.functions.getTokenVestingIds(
            Web3.to_checksum_address(token), start_index, end_index
        ).call()

    def get_vesting_details_batch(self, vesting_ids: list):
        """Returns vesting details for multiple vesting IDs in a single call."""
        return self.contract.functions.getVestingDetailsBatch(vesting_ids).call()

    def get_vesting_count(self) -> int:
        """Returns the total number of vesting schedules created."""
        return self.contract.functions.vestingCount().call()

    def get_vestings_by_beneficiary(self, beneficiary: str) -> list:
        """Returns all vesting IDs for a beneficiary."""
        return self.contract.functions.getVestingsByBeneficiary(Web3.to_checksum_address(beneficiary)).call()

    def get_vestings_by_creator(self, creator: str) -> list:
        """Returns all vesting IDs for a creator."""
        return self.contract.functions.getVestingsByCreator(Web3.to_checksum_address(creator)).call()
